### ISSUE TITLE
Low precision integers support

### ISSUE THREAD

**Issue by soupault** [blink1073/tifffile#42]
Hi!
`README.rst` states that:
>Only a subset of the TIFF specification is supported, mainly uncompressed and losslessly compressed 2**(0 to 6) bit integer, 16, 32 and 64-bit float, grayscale and RGB(A) images, which are commonly used in bio-scientific imaging.

>2**(0 to 6) bit integer

Is there really a way to write boolean arrays to .tiff from Python? Trying with np.bool_ doesn't work in my environment.

**Comment by cgohlke** [comment]
It's not clear, but those statements belong to the reading capabilities. Tifffile can not currently write packed integers.

**Comment by soupault** [comment]
Ah, I see. Thanks for the clarification!
Please, feel free to close the issue if you don't anticipate this happening in the next couple of years :).

**Comment by cgohlke** [comment]
Is the use case just for 1 bit (bool) or any low bit integers? Bool should be relatively easy to add via numpy.packbits.

**Comment by soupault** [comment]
Yes, thanks for the hint! In my case it is indeed a 1-bit data. Although, I'm curious how fast `[un]packbits` is...

**Comment by cgohlke** [comment]
I think support for bool arrays would be a good addition. If you are very concerned about speed, better use one of the C libraries such as libtiff, GDAL, OpenImageIO, GraphicsMagick. Implemented mostly in Python, tifffile is only fast in a few cases such as large contiguous data that can be memory mapped to numpy arrays. Otherwise the overhead of Python is significant; detrimental for files with huge number of tiles, pages, or non-contiguous strips. The main advantages of tifffile are the straightforward way to write nd numpy arrays to TIFF and read them back. Tifffile can also handle a few TIFF like formats such as ImageJ, LSM, SGI, and OME, that other libraries struggle with.

**Comment by soupault** [comment]
>If you are very concerned about speed, better use one of the C libraries such as libtiff, GDAL, OpenImageIO, GraphicsMagick. Implemented mostly in Python, tifffile is only fast in a few cases such as large contiguous data that can be memory mapped to numpy arrays. Otherwise the overhead of Python is significant; detrimental for files with huge number of tiles, pages, or non-contiguous strips. The main advantages of tifffile are the straightforward way to write nd numpy arrays to TIFF and read them back. Tifffile can also handle a few TIFF like formats such as ImageJ, LSM, SGI, and OME, that other libraries struggle with.

I see. I've already some experience with GDAL - it is much harder to install. :)

>I think support for bool arrays would be a good addition. 

How do you see this implemented? Should there be a dedicated tiff flag to distinguish packed arrays from unpacked?

**Comment by cgohlke** [comment]
> Should there be a dedicated tiff flag to distinguish packed arrays from unpacked?

I already implemented this. Numpy arrays of dtype `bool` are packed using numpy.packbits and written to a bilevel TIFF: PhotometricInterpretation=MINISWHITE, no SampleFormat, no BitsPerSample (TIFF defaults to 1 bit), no PlanarConfiguration. I do not plan to implement packing for other bit depths, tiling, or compression.

**Comment by soupault** [comment]
@cgohlke am I doing something wrong? :
```
In [19]: a = np.random.randn(100000).reshape(10, 100, 100)

In [20]: a = (a > 0).astype(np.bool_)

In [21]: a.shape
Out[21]: (10, 100, 100)

In [22]: a.dtype
Out[22]: dtype('bool')

In [23]: tifffile.imsave('exp.tiff', a)
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-23-86ec4d4a9c5a> in <module>()
----> 1 tifffile.imsave('exp.tiff', a)

~/Workspace/project/py3/lib/python3.6/site-packages/tifffile/tifffile.py in imsave(file, data, shape, dtype, bigsize, **kwargs)
    496 
    497     with TiffWriter(file, **tifargs) as tif:
--> 498         return tif.save(data, shape, dtype, **kwargs)
    499 
    500 

~/Workspace/project/py3/lib/python3.6/site-packages/tifffile/tifffile.py in save(self, data, shape, dtype, returnoffset, photometric, planarconfig, tile, contiguous, align, truncate, compress, rowsperstrip, predictor, colormap, description, datetime, resolution, metadata, extratags)
   1239                 addtag('TileDepth', 'I', 1, tile[0])
   1240         addtag('NewSubfileType', 'I', 1, 0)
-> 1241         sampleformat = {'u': 1, 'i': 2, 'f': 3, 'c': 6}[datadtype.kind]
   1242         addtag('SampleFormat', 'H', samplesperpixel,
   1243                (sampleformat,) * samplesperpixel)

KeyError: 'b'
```

**Comment by cgohlke** [comment]
You'll need [tiffffile > 2018.6.15](https://www.lfd.uci.edu/~gohlke/code/tifffile.py.html). This repository has not been updated yet.

**Comment by blink1073** [comment]
v0.15 is now on PyPI, conda-forge pending
